File: problems/71/71.simplify-path.py
#
# @lc app=leetcode id=71 lang=python3
#
# [71] Simplify Path
#


# @lc code=start
class Solution:
    def simplifyPath(self, path: str) -> str:
        # convert absolute path to canonical path
        # canonical path
        # 1. path starts with single "/"
        # 2. any two directories are separated by a single slash '/'.
        # 3. path does not end with trailing "/"
        # 4. path only contains the directories on the path from the root directory to the target file or directory
        #    (i.e., no period '.' or double period '..')
        # ie:
        # path = "/home/"
        # -> "/home"
        # path = "/../"
        # -> "/"
        # path = "/home//foo/"
        # -> "/home/foo"
        #
        # approach:
        # stack -> keep track of directory
        # continue if /
        # if /./ -> do nth
        # if /../ -> pop stack
        # if /..dir/ -> append stack

        path += "/"
        stack = []
        curr = ""
        for c in path:
            if c == "/":
                if not curr:
                    continue
                if curr == ".." and stack:
                    stack.pop()
                if curr != ".." and curr != ".":
                    stack.append(curr)
                curr = ""
            else:
                curr += c

        if not stack:
            return "/"

        res = ""
        for val in stack:
            res += "/" + val

        return res


# @lc code=end

File: problems/71/71.simplify-path.go
/*
 * @lc app=leetcode id=71 lang=golang
 *
 * [71] Simplify Path
 */

package solution

// @lc code=start
func simplifyPath(path string) string {
	// O(N) time, O(N) space
	// stack to keep track of directories

	path += "/" // to make sure all paths end with /
	n, stack := len(path), make([]string, 0)
	curr := ""

	for i := 0; i < n; i++ {
		c := string(path[i])
		if c == "/" {
			if curr == "" {
				continue
			} else if curr == ".." && len(stack) > 0 {
				stack = stack[:len(stack)-1]
			} else if curr != ".." && curr != "." {
				stack = append(stack, curr)
			}
			curr = ""
		} else {
			curr += c
		}
	}

	if len(stack) == 0 {
		return "/"
	}

	res := ""
	for _, dir := range stack {
		res += "/" + dir
	}

	return res
}

// @lc code=end
